c_src/filewatch_inotify.c

```diff
@@ -181,41 +185,65 @@ static void send_output(ErlDrvData self_)
 {
     struct instance *self = (struct instance *)self_;
 
+    size_t len = notif_set_get_length(self->notif_set);
+
     size_t pos = 0;
-    ErlDrvTermData buf[(MAX_EVENTS * EVENT_TUPLE_LENGTH) + BUFFER_TAIL_SIZE];
+    ErlDrvTermData buf[(len * EVENT_TUPLE_LENGTH) + BUFFER_TAIL_SIZE];
 
     buf[pos++] = ERL_DRV_PORT;
     buf[pos++] = driver_mk_port(self->port);
 
-    struct message *msg;
-    for(uint i = 0; i < self->n_events; ++i) {
-        msg = &(self->msgs[i]);
+    struct Message msg[len];
+    for(size_t i = 0; i < len; i++) {
+        msg[i].name = driver_alloc(NAME_MAX);
+    }
+    int num_entries_removed = 0;
+    struct Notif_set_itr *itr = notif_set_itr_init(self->notif_set);
+    if(itr == NULL) {
+        stop(self_);
+    }
+    while(notif_set_itr_next(itr, &msg[num_entries_removed])) {
         buf[pos++] = ERL_DRV_INT;
-        buf[pos++] = msg->wd;
+        buf[pos++] = msg[num_entries_removed].wd;
         buf[pos++] = ERL_DRV_STRING;
-        buf[pos++] = (ErlDrvTermData) msg->name;
-        buf[pos++] = strnlen(msg->name, NAME_MAX);
+        buf[pos++] = (ErlDrvTermData) msg[num_entries_removed].name;
+        buf[pos++] = strnlen(msg[num_entries_removed].name, NAME_MAX);
         buf[pos++] = ERL_DRV_TUPLE;
         buf[pos++] = BUFFER_TUPLE_SIZE;
+        
```

Comment:
Weird that GH isn't picking up this trailing whitespace... It's usually pretty good about this.

Make sure you `:%s/\s\+$//e` at some point :smile_cat: